### PR TITLE
fix: Do not create pods under shutdown strategy

### PR DESCRIPTION
**Pull request by simster7**
Fixes https://github.com/argoproj/argo/issues/5052

### CODE REVIEW

File: workflow/controller/workflowpod.go [hunk 1]
```diff
@@ -132,6 +132,12 @@ type createWorkflowPodOpts struct {
 func (woc *wfOperationCtx) createWorkflowPod(ctx context.Context, nodeName string, mainCtr apiv1.Container, tmpl *wfv1.Template, opts *createWorkflowPodOpts) (*apiv1.Pod, error) {
 	nodeID := woc.wf.NodeID(nodeName)
 
+	if !woc.execWf.Spec.Shutdown.ShouldExecute(opts.onExitPod) {
+		// Do not create pods if we are shutting down
+		woc.markNodePhase(nodeName, wfv1.NodeFailed, fmt.Sprintf("workflow shutdown with strategy:  %s", woc.execWf.Spec.Shutdown))
```
Comment (alexec) [inline workflow/controller/workflowpod.go:137]: why not `NodeSkip`?

Comment (simster7) [reply workflow/controller/workflowpod.go:137]: 🤔... great suggestion!

Comment (alexec) [reply workflow/controller/workflowpod.go:137]: I think you could write an e2e test for this

Comment (alexec) [reply workflow/controller/workflowpod.go:137]: signals_test.go

Comment (simster7) [reply workflow/controller/workflowpod.go]: Will do

Comment (simster7) [reply workflow/controller/workflowpod.go]: Done

**Comment by simster7** [comment]
@alexec Need a review on this

File: workflow/controller/workflowpod.go [hunk 2]
```diff
@@ -147,6 +147,12 @@ func (woc *wfOperationCtx) createWorkflowPod(ctx context.Context, nodeName strin
 		}
 	}
 
+	if !woc.execWf.Spec.Shutdown.ShouldExecute(opts.onExitPod) {
```
[the remark this answers is not in the document]
Comment (simster7) [reply workflow/controller/workflowpod.go:150]: Tests already exist:

https://github.com/argoproj/argo-workflows/blob/57b38282556e0e6ad00b431845ac0168c7f5b6e4/pkg/apis/workflow/v1alpha1/workflow_types_test.go#L457-L462

File: test/e2e/signals_test.go [hunk 3]
```diff
@@ -78,6 +78,29 @@ func (s *SignalsSuite) TestTerminateBehavior() {
 		})
 }
 
+func (s *SignalsSuite) TestDoNotCreatePodsUnderStopBehavior() {
```
[the remark this answers is not in the document]
Comment (simster7) [reply test/e2e/signals_test.go:81]: Added

File: test/e2e/functional/stop-terminate-2.yaml [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  entrypoint: diamond
+  templates:
+    - name: diamond
+      steps:
+        - - name: A
+            template: echo
+        - - name: B
+            template: echo-2
+
+    - name: echo
```
[the remark this answers is not in the document]
Comment (simster7) [reply test/e2e/functional/stop-terminate-2.yaml:17]: Changed

File: test/e2e/functional/stop-terminate.yaml [hunk 5]
```diff
@@ -18,8 +18,8 @@ spec:
     - name: echo
       container:
         image: argoproj/argosay:v2
```
[the remark this answers is not in the document]
Comment (simster7) [reply test/e2e/functional/stop-terminate.yaml:20]: Switched

**Comment by simster7** [comment]
@alexec Addressed comments, please re-review

File: test/e2e/functional/stop-terminate-2.yaml [hunk 6]
```diff
@@ -0,0 +1,24 @@
+apiVersion: argoproj.io/v1alpha1
+kind: Workflow
+metadata:
+  generateName: stop-terminate-
+  labels:
+    argo-e2e: true
+spec:
+  entrypoint: diamond
```
[the remark this answers is not in the document]
Comment (simster7) [reply test/e2e/functional/stop-terminate-2.yaml]: Changed

File: test/e2e/signals_test.go [hunk 7]
```diff
@@ -78,6 +78,29 @@ func (s *SignalsSuite) TestTerminateBehavior() {
...
+	s.Given().
+		Workflow("@functional/stop-terminate-2.yaml").
+		When().
+		SubmitWorkflow().
+		WaitForWorkflow(fixtures.ToStart, "to start").
+		RunCli([]string{"stop", "@latest"}, func(t *testing.T, output string, err error) {
+			assert.NoError(t, err)
+			assert.Regexp(t, "workflow stop-terminate-.* stopped", output)
+		}).
+		WaitForWorkflow().
```
[the remark this answers is not in the document]
Comment (simster7) [reply test/e2e/signals_test.go]: Increased

File: test/e2e/signals_test.go [hunk 8]
```diff
@@ -78,6 +78,30 @@ func (s *SignalsSuite) TestTerminateBehavior() {
...
+	s.Given().
+		Workflow("@functional/stop-terminate-2.yaml").
+		When().
+		SubmitWorkflow().
+		WaitForWorkflow(fixtures.ToStart, "to start").
+		RunCli([]string{"stop", "@latest"}, func(t *testing.T, output string, err error) {
+			assert.NoError(t, err)
+			assert.Regexp(t, "workflow stop-terminate-.* stopped", output)
+		}).
+		WaitForWorkflow(45 * time.Second).
```
Comment (alexec) [inline test/e2e/signals_test.go:92]: can we make this 1m?

File: test/e2e/functional/stop-terminate.yaml (hunk 5 above)
Comment (alexec) [inline test/e2e/functional/stop-terminate.yaml:20]: can we make this v1?

File: test/e2e/functional/stop-terminate-2.yaml [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+    - name: main
+      steps:
+        - - name: A
+            template: sleep
+        - - name: B
+            template: pass
+
+    - name: sleep
+      container:
+        image: argoproj/argosay:v2
```
Comment (alexec) [inline test/e2e/functional/stop-terminate-2.yaml:19]: can we make this v1? I think you need to change sleep to remove "s" suffixe

**Review by alexec — APPROVED** [review]
some minor comments

**Pull request merged**